File: Difficulty: Medium/Tree Boundary Traversal/tree-boundary-traversal.cpp
/*
// Tree Node
class Node {
  public:
    int data;
    Node* left;
    Node* right;

    // Constructor to initialize a new node
    Node(int val) {
        data = val;
        left = NULL;
        right = NULL;
    }
};
*/

class Solution {
    
  public:
  bool isleaf(Node* root) {
        return (root->left == NULL && root->right == NULL);
    }
  void addleftnodes(Node* root,vector<int>& result){
        Node* cur =root->left;
        while (cur){
            if(!isleaf(cur)) result.push_back(cur->data);
            if(cur->left) cur=cur->left;
            else cur=cur->right;
        }

    }
    
    void addleafnodes(Node* root,vector<int>& result){
        if(isleaf(root)){
            result.push_back(root->data);
            return;
        }
        if(root->left) addleafnodes(root->left,result);
        if(root->right) addleafnodes(root->right,result);
    }
    
    void addrightnodes(Node* root,vector<int>& result){
        Node* cur = root->right;
        vector<int>temp;
        while(cur){
            if(!isleaf(cur)) temp.push_back(cur->data);
            if(cur->right) cur=cur->right;
            else cur=cur->left;
        }
        for(int i=temp.size()-1;i>=0;i--){
            result.push_back(temp[i]);
        }
    }
    vector<int> boundaryTraversal(Node *root) {
        // code here
        vector<int> result;
        if(root==NULL) return result;
        if(!isleaf(root)) result.push_back(root->data);
        
        addleftnodes(root,result);
        addleafnodes(root,result);
        addrightnodes(root,result);
        
        return result;
        
    }
    
  
};

File: Difficulty: Medium/Tree Boundary Traversal/tree-boundary-traversal.py
'''
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None
'''
class Solution:
    def isleafnode(self,root):
            return root.left==None and root.right==None
            
    def leftboun(self,root,res):
        curr=root.left
        while curr:
            if not self.isleafnode(curr):
                res.append(curr.data)
            
            if curr.left:
                curr=curr.left
            else:
                curr=curr.right
    
    def rightboun(self,root,res):
        curr=root.right
        temp=[]
        while curr:
            if not self.isleafnode(curr):
                temp.append(curr.data)
            if curr.right:
                curr=curr.right
            else:
                curr=curr.left
                
        
        for i in range(len(temp)-1,-1,-1):
            res.append(temp[i])
        
    def addleafs(self,root,res):
        if self.isleafnode(root):
            res.append(root.data)
        
        if root.left:
            self.addleafs(root.left,res)
        
        if root.right:
            self.addleafs(root.right,res)
    
    
    def boundaryTraversal(self, root):
        # Code here
        res=[]
        if not root:
            return res
        
        if not self.isleafnode(root):
            res.append(root.data)
        
        self.leftboun(root,res)
        self.addleafs(root,res)
        self.rightboun(root,res)
        
        return res
        
        
            
                